sys/include/auto_init_macros.h

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (C) 2021 Otto-von-Guericke Universität Magdebug
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_auto_init_macros Macros to setup the auto-initialization sequence of modules
+ * @ingroup     sys
+ * @brief       Auto-initialization macros
+ *
+ * @{
+ * @file
+ * @brief       Auto-initialization macros
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ */
+#ifndef AUTO_INIT_MACROS_H
+#define AUTO_INIT_MACROS_H
+
+#include "debug.h"
+#include "xfa.h"
+
+#ifdef __cpluspluc
+extern "C" {
+#endif
+
+#if !defined(CONCAT) || defined(DOXYGEN)
+/**
+ * @name    Classic concat macro
+ * @internal
+ * @{
+ */
+#define CONCAT_HELP(a, b) a ## b
+#define CONCAT(a, b) CONCAT_HELP(a, b)
+/** @} */
+#endif
+
+/**
+ * @brief   Auto init function type
+ */
+typedef void(*auto_init_fn)(void);
+
+/**
+ * @name    Priority levels of auto-initialized modules from highest to lowest
+ *
+ * If you have a custom module that needs to be initialized after some other module,
+ * for example after `ztimer`, you can add another priority level like that:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ * #include "auto_init_macros.h"
+ *
+ * #define AUTO_INIT_PRIO_MY_MODULE AUTO_INIT_PRIO_AFTER(AUTO_INIT_PRIO_ZTIMER, 0)
+ * AUTO_INIT_MODULE(my_module, AUTO_INIT_PRIO_MY_MODULE, my_module_init);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * @ref AUTO_INIT_MODULE
+ * @internal
+ *
+ * @experimental
+ * @{
+ */
+#define AUTO_INIT_PRIO_0                            a_
```

Comment:
why is this needed? the symbols get sorted alphabetically, but if we just add a fairly large constant (like, 10000000), can't we use plain numeric values?

Reply:
I think using numbers as priority prefix could be messy, because e.g. `10000010 < 10000002` in alphanumeric order. That´s why I think it is good to abstract priority levels with `AUTO_INIT_PRIO_0 ` to  `AUTO_INIT_PRIO_n`. If you need to compare priorities, you could compare the `CONFIG_` values. That would be needed if your custom modules depends on more then one module and you need to choose the one with the lowest priority to depend on. Or do you have something else in mind?

Reply:
Yeah that bites me not for the first time.
I wonder if we can use `SORT_BY_INIT_PRIORITY` in the linker script for this.

Reply:
Ah what am I saying '10 < 2' but `10000010` > `10000002`.

Reply:
doesn't seem like, unless we want to use "init_array" or similar. here's the sort_by_init_priority logic of binutils:

```
   393   if (strncmp (name, ".init_array.", 12) == 0                                                                                               
   394       || strncmp (name, ".fini_array.", 12) == 0)                                                                                           
   395     {                                                                                                                                       
   396       init_priority = strtoul (name + 12, &end, 10);                                                                                        
   397       return *end ? 0 : init_priority;                                                                                                      
   398     }                                                                                                                                       
   399   else if (strncmp (name, ".ctors.", 7) == 0                                                                                                
   400 ›      || strncmp (name, ".dtors.", 7) == 0)                                                                                                
   401     {                                                                                                                                       
   402       init_priority = strtoul (name + 7, &end, 10);                                                                                         
   403       return *end ? 0 : 65535 - init_priority;                                                                                              
   404     }                                                                                                                                       
```

Reply:
> Ah what am I saying `10` < `2` but `10000010` > `10000002`

ah, I got confused again. that's why adding a large sonstant helps, because alphabetically, `102` gets correctly sorted before `110`, but `2` would be sorted after `10`. numerically, both are sorted correctly.

Reply:
but, still we cannot do arithmetic. I hoped that we could express a priority like `(1000000 | DEP1_PRIO | DEP2_PRIO ) + 1`, but the preprocessor would just pass that through to the symbol name. :(